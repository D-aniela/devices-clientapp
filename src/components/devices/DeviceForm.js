import React from "react";
import PropTypes from "prop-types";
import TextInput from "../common/TextInput";
import SelectInput from "../common/SelectInput";

const DeviceForm = ({
  device,
  typeDevice,
  onSave,
  onChange,
  saving = false,
  errors = {},
}) => {
  return (
    <form onSubmit={onSave}>
      <h2>{device.id ? "Edit" : "Add"} Device</h2>
      {errors.onSave && (
        <div className="alert alert-danger" role="alert">
          {errors.onSave}
        </div>
      )}

      <TextInput
        name="system_name"
        label="System Name"
        value={device.system_name}
        onChange={onChange}
        error={errors.system_name}
      />

      <SelectInput
        name="type"
        label="Device Type"
        value={device.type}
        defaultOption="Select Device Type"
        options={typeDevice.map((types) => ({
          value: types.type,
          text: types.name,
        }))}
        onChange={onChange}
        error={errors.type}
      />

      <TextInput
        name="hdd_capacity"
        label="Capacity"
        value={device.hdd_capacity}
        onChange={onChange}
        error={errors.hdd_capacity}
      />

      <button type="submit" disabled={saving} className="btn btn-primary">
        {saving ? "Saving..." : "Save"}
      </button>
    </form>
  );
};

DeviceForm.propTypes = {
  device: PropTypes.object.isRequired,
  typeDevice: PropTypes.array.isRequired,
  errors: PropTypes.object,
  onSave: PropTypes.func.isRequired,
  onChange: PropTypes.func.isRequired,
  saving: PropTypes.bool,
};

export default DeviceForm;
