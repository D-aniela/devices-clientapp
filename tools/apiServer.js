/* eslint-disable no-console */
const jsonServer = require("json-server");
const server = jsonServer.create();
const path = require("path");
const router = jsonServer.router(path.join(__dirname, "devices.json"));

// Can pass a limited number of options to this to override (some)
const middlewares = jsonServer.defaults({
  // Display json-server's built in homepage when json-server starts.
  static: "node_modules/json-server/dist",
});

// Set default middlewares (logger, static, cors and no-cache)
server.use(middlewares);

// To handle POST, PUT and PATCH you need to use a body-parser. Using JSON Server's bodyParser
server.use(jsonServer.bodyParser);

// Simulate delay on all requests
server.use(function (req, res, next) {
  setTimeout(next, 0);
});

// Declaring custom routes below. Add custom routes before JSON Server router

// Add createdAt to all POSTS
server.use((req, res, next) => {
  if (req.method === "POST") {
    req.body.createdAt = Date.now();
  }
  // Continue to JSON Server router
  next();
});

server.post("//", function (req, res, next) {
  const error = validateDevice(req.body);
  if (error) {
    res.status(400).send(error);
  } else {
    console.log(req.body);
    req.body.system_name = createSlug(req.body.system_name); // Generate a system name url for new devices
    next();
  }
});

// Use default router
server.use(router);

// Start server
const port = 3001;
server.listen(port, () => {
  console.log(`JSON Server is running on port ${port}`);
});

// Returns a URL friendly slug
function createSlug(value) {
  return value
    .replace(/[^a-z0-9_]+/gi, "-")
    .replace(/^-|-$/g, "")
    .toLowerCase();
}

function validateDevice(device) {
  if (!device.system_name) return "Device name is required.";
  if (!device.type) return "Device type is required.";
  if (!device.hdd_capacity) return "Device capacity is required.";
  if (isNaN(device.hdd_capacity)) return "Must input numbers";
  return "";
}
