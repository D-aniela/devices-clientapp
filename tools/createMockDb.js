/* eslint-disable no-console */
const fs = require("fs");
const path = require("path");
const mockData = require("./mockData");

const { devices, typeDevice } = mockData;
const data = JSON.stringify({ devices, typeDevice });
const filepath = path.join(__dirname, "devices.json");

fs.writeFile(filepath, data, function(err) {
  err ? console.log(err) : console.log("Mock DB created.");
});
