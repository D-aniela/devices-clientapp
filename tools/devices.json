{
  "devices": [
    {
      "id": "e8okoP2l5",
      "system_name": "DESKTOP-SMART",
      "type": "WINDOWS_WORKSTATION",
      "hdd_capacity": "10"
    },
    {
      "id": "Th3ngERn9",
      "system_name": "MAC-LEADER",
      "type": "MAC",
      "hdd_capacity": "2048"
    },
    {
      "id": "Q1JdBnE12",
      "system_name": "ARMANDO-SERVER",
      "type": "WINDOWS_SERVER",
      "hdd_capacity": "256"
    },
    {
      "id": "e7ocoQ2n3",
      "system_name": "MIGUEL-PC",
      "type": "WINDOWS_WORKSTATION",
      "hdd_capacity": "500"
    },
    {
      "id": "Jj5bn3G2H",
      "system_name": "FIRST-MAC",
      "type": "MAC",
      "hdd_capacity": "180"
    },
    {
      "id": "GT556QGnk",
      "system_name": "GOOD-MAC",
      "type": "MAC",
      "hdd_capacity": "500"
    },
    {
      "id": "R5LdSnQhY",
      "system_name": "SERVER-ONE",
      "type": "WINDOWS_SERVER",
      "hdd_capacity": "50"
    },
    {
      "id": "ab1coL2n9",
      "system_name": "GILBERT-COMPUTER",
      "type": "WINDOWS_WORKSTATION",
      "hdd_capacity": "750"
    },
    {
      "id": "LM5dBnJ2G",
      "system_name": "MOON-SMART",
      "type": "WINDOWS_SERVER",
      "hdd_capacity": "256"
    },
    {
      "id": "Up5bcEQp4",
      "system_name": "JULIO-MAC-LOCAL",
      "type": "MAC",
      "hdd_capacity": "512"
    },
    {
      "id": "DRaETcR",
      "system_name": "hola",
      "type": "MAC",
      "hdd_capacity": "256",
      "createdAt": 1615537938719
    },
    {
      "id": "R4JjV9r",
      "system_name": "segundo",
      "type": "WINDOWS_SERVER",
      "hdd_capacity": "258",
      "createdAt": 1615539462208
    },
    {
      "id": "cPkdPC0",
      "system_name": "2",
      "type": "MAC",
      "hdd_capacity": "256",
      "createdAt": 1615539664707
    }
  ],
  "typeDevice": [
    {
      "type": "MAC",
      "name": "MAC"
    },
    {
      "type": "WINDOWS_SERVER",
      "name": "Windows Server"
    },
    {
      "type": "WINDOWS_WORKSTATION",
      "name": "Windows Workstation"
    }
  ]
}