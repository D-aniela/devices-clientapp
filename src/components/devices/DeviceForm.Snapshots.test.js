import React from "react";
import DeviceForm from "./DeviceForm";
import renderer from "react-test-renderer";
import { devices, typeDevice } from "./../../../tools/mockData";

it("sets submit button label 'Saving...' when saving is true", () => {
  const tree = renderer.create(
    <DeviceForm
      device={devices[0]}
      typeDevice={typeDevice}
      onSave={jest.fn()}
      onChange={jest.fn()}
      saving
    />
  );

  expect(tree).toMatchSnapshot();
});

it("sets submit button label 'Save' when saving is false", () => {
  const tree = renderer.create(
    <DeviceForm
      device={devices[0]}
      typeDevice={typeDevice}
      onSave={jest.fn()}
      onChange={jest.fn()}
      saving={false}
    />
  );

  expect(tree).toMatchSnapshot();
});
