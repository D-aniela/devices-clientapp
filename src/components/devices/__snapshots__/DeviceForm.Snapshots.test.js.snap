// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sets submit button label 'Save' when saving is false 1`] = `
<form
  onSubmit={[MockFunction]}
>
  <div
    className="bgColor"
  />
  <h2
    className="edit_add"
  >
    Edit
     Device
  </h2>
  <div
    className="form-group"
  >
    <label
      htmlFor="system_name"
    >
      System Name *
    </label>
    <div
      className="field"
    >
      <input
        className="form-control"
        name="system_name"
        onChange={[MockFunction]}
        type="text"
        value="DESKTOP-SMART"
      />
    </div>
  </div>
  <div
    className="form-group"
  >
    <label
      htmlFor="type"
    >
      Device Type *
    </label>
    <div
      className="field"
    >
      <select
        className="form-control"
        name="type"
        onChange={[MockFunction]}
        value="WINDOWS_WORKSTATION"
      >
        <option
          value=""
        >
          Select Device Type
        </option>
        <option
          value="MAC"
        >
          MAC
        </option>
        <option
          value="WINDOWS_SERVER"
        >
          Windows Server
        </option>
        <option
          value="WINDOWS_WORKSTATION"
        >
          Windows Workstation
        </option>
      </select>
    </div>
  </div>
  <div
    className="form-group"
  >
    <label
      htmlFor="hdd_capacity"
    >
      Capacity *
    </label>
    <div
      className="field"
    >
      <input
        className="form-control"
        name="hdd_capacity"
        onChange={[MockFunction]}
        type="text"
        value="10"
      />
    </div>
  </div>
  <button
    className="btn btn-primary"
    disabled={false}
    type="submit"
  >
    Save
  </button>
</form>
`;

exports[`sets submit button label 'Saving...' when saving is true 1`] = `
<form
  onSubmit={[MockFunction]}
>
  <div
    className="bgColor"
  />
  <h2
    className="edit_add"
  >
    Edit
     Device
  </h2>
  <div
    className="form-group"
  >
    <label
      htmlFor="system_name"
    >
      System Name *
    </label>
    <div
      className="field"
    >
      <input
        className="form-control"
        name="system_name"
        onChange={[MockFunction]}
        type="text"
        value="DESKTOP-SMART"
      />
    </div>
  </div>
  <div
    className="form-group"
  >
    <label
      htmlFor="type"
    >
      Device Type *
    </label>
    <div
      className="field"
    >
      <select
        className="form-control"
        name="type"
        onChange={[MockFunction]}
        value="WINDOWS_WORKSTATION"
      >
        <option
          value=""
        >
          Select Device Type
        </option>
        <option
          value="MAC"
        >
          MAC
        </option>
        <option
          value="WINDOWS_SERVER"
        >
          Windows Server
        </option>
        <option
          value="WINDOWS_WORKSTATION"
        >
          Windows Workstation
        </option>
      </select>
    </div>
  </div>
  <div
    className="form-group"
  >
    <label
      htmlFor="hdd_capacity"
    >
      Capacity *
    </label>
    <div
      className="field"
    >
      <input
        className="form-control"
        name="hdd_capacity"
        onChange={[MockFunction]}
        type="text"
        value="10"
      />
    </div>
  </div>
  <button
    className="btn btn-primary"
    disabled={true}
    type="submit"
  >
    Saving...
  </button>
</form>
`;
